File: RandomVariablesExpectation/Coupon.py
'''
Vazquez Choreno Luis Ernesto
Randomized Algorithms
Application: Example Coupon Collectors Problem


python3 Coupon.py
'''

from random import seed
from random import randint
import matplotlib.pyplot as plt
import numpy as np
def get_rng(value):
    return np.random.randint(low=1, high=value+1)



def randomized_algorithm(num_coupons):
    collector  = []
    for i in range(0,num_coupons+1):
        collector.append(False)
    counter = 0
    iteration = 0
    while counter < num_coupons:
        iteration += 1
        rng = int(get_rng(num_coupons))
        if collector[rng] == True:
            continue
        else:
            collector[rng] = True
            counter+=1
    return iteration



def save_plot(x,expectedY,practiceY,name):
    plt.plot(x,expectedY, label='Valor esperado')
    plt.plot(x,practiceY, label ='Algoritmo Aleatorio')
    plt.legend(framealpha=1, frameon=True);
    plt.ylabel("Intentos hasta obtener c cupones")
    plt.xlabel('c cupones')
    #plt.savefig(name)
    plt.show()

def main():
    num_coupons = 100

    print("Expected value: " +  str(num_coupons * np.log(num_coupons) + num_coupons))
    print("Practice " + str(randomized_algorithm(num_coupons)))

    x = []
    expectedY = []
    practiceY = []
    for i in range(1,num_coupons + 1,5):
        expectedY.append(i * np.log(i) + i)
        practiceY.append(randomized_algorithm(i))
        x.append(i)
    save_plot(x,expectedY,practiceY,"Plots/PythonFile.png")

main()

File: RandomVariablesExpectation/Coupon.cpp
/*
Vázquez Choreño Luis Ernesto
Randomized Algorithms
Application: Example Coupon Collectors Problem



g++ -std=c++11 Coupon.cpp  && ./a.out

*/

#include <bits/stdc++.h>

using namespace std;
typedef long long lli;



int get_rng(int value){ return (rand()%value ) + 1;}

int randomized_algorithm(int num_coupons){
  vector<bool> collector(num_coupons + 5,false);
  int counter = 0;
  int iteration = 0;
  while(counter < num_coupons){
    iteration++;
    int rng = get_rng(num_coupons);
    if(collector[rng] == true){
      continue;
    }else{
      collector[rng] = true;
      counter++;
    }
  }
  return iteration;
}






int main(){
  srand(time(NULL));
  int num_coupons = 100;
  cout << "Expected value: "  << ((double)(num_coupons) * log(num_coupons) + num_coupons) << "\n";
  cout << "Practice: " << randomized_algorithm(num_coupons) << "\n";

  vector<int> x;
  vector<double> expected_y;
  vector<double> practice_y;

  for(int i=1;i<=num_coupons;i+=5){
    x.push_back(i);
    expected_y.push_back( i * log(i) + i);
    practice_y.push_back(randomized_algorithm(i));
  }


}
